File: android/src/main/java/com/plugins/forgerockbridge/OTPTokenHandler.java
package com.plugins.forgerockbridge;

import android.content.Context;
import android.util.Log;


import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;

import org.forgerock.android.auth.Account;
import org.forgerock.android.auth.FRAClient;
import org.forgerock.android.auth.FRListener;
import org.forgerock.android.auth.FRSession;
import org.forgerock.android.auth.FRUser;
import org.forgerock.android.auth.Logger;
import org.forgerock.android.auth.Mechanism;
import org.forgerock.android.auth.NodeListener;
import org.forgerock.android.auth.OathMechanism;
import org.forgerock.android.auth.OathTokenCode;
import org.forgerock.android.auth.UserInfo;
import org.forgerock.android.auth.exception.AuthenticatorException;
import org.json.JSONObject;

import com.plugins.forgerockbridge.ErrorHandler.FRException;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Objects;

import okhttp3.Call;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;

import org.json.*;

public class OTPTokenHandler {
    private static final String TAG = "ForgeRockBridge";
    private static final MediaType MEDIA_JSON = MediaType.get("application/json; charset=utf-8");

    private static OkHttpClient httpClient() {
        HttpLoggingInterceptor httpLog = new HttpLoggingInterceptor(msg -> Log.d(TAG, msg));
        httpLog.setLevel(HttpLoggingInterceptor.Level.BODY);
        return new OkHttpClient.Builder()
                .addInterceptor(httpLog)
                .followRedirects(false)
                .build();
    }

  public static void startOtpJourney(PluginCall call, Context context, NodeListener<FRSession> listener ) {
        String journey = call.getString("journey");

        if (journey == null) {
            Log.e(TAG, "[OTPTokenHandler]: Error in get Journey");
          ErrorHandler.reject(call, ErrorHandler.ErrorCode.MISSING_JOURNEY);
            return;
        }
      Log.d(TAG, "[OTPTokenHandler]: PASO AQUI "+journey);

      FRSession.authenticate(context, journey, listener);

    }

  public static void hasRegisteredMechanism(PluginCall call, Context context) {
        try {
            Log.d(TAG, "[OTPTokenHandler]: Sending existence of mechanism true or false");
            JSObject response = validateExistMechanism(context);
            boolean emptyMechanism = response.getBool("empty");

            JSObject result = new JSObject();

            result.put("empty", emptyMechanism);

            call.resolve(result);
        } catch (Exception e) {
          Log.e(TAG, "[OTPTokenHandler]: RETURN FRE026 ERROR from onException");
          ErrorHandler.reject(call, ErrorHandler.ErrorCode.NO_ACCOUNTS_REGISTERED);
        }
  }

  public static void validateExistenceOTP(PluginCall call, Context context) {
        FRUser user = FRUser.getCurrentUser();
        if (user == null) {
            Log.e(TAG, "[OTPTokenHandler]: FRUser is null");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.GETTING_USER_INFO);
        }

        user.getUserInfo(new FRListener<UserInfo>() {
            @Override
            public void onSuccess(UserInfo userInfo) {
                String uuid = userInfo.getSub();
                checkServerAndDeviceOtpState(call, context, uuid);

            }

            @Override
            public void onException(Exception e) {
                Log.e(TAG, "[OTPTokenHandler]: Error getting userInfo", e);
                ErrorHandler.reject(call, ErrorHandler.ErrorCode.GETTING_USER_INFO);
            }
        });
  }

  public static void generateOTP(PluginCall call, Context context) {
        try {
            Log.d(TAG, "[OTPTokenHandler]: Generate OTP with OathTokenCode");

            FRAClient fraClient = initClient(context);

            Account account = getAccount(fraClient);
            OathMechanism mechanism = getOathMechanism(account);
            OathTokenCode token = mechanism.getOathTokenCode();
            String otp = token.getCurrentCode();

            var expiresIn = getRemainingTime(token);


            JSObject result = new JSObject();
            result.put("otp", otp);
            result.put("expiresIn", expiresIn);

            call.resolve(result);

        } catch (FRException e) {
            Log.e(TAG, "[OTPTokenHandler]: RETURN ERROR in get OTP with FRException");
          ErrorHandler.reject(call, e.getCode());
        } catch (Exception e) {
            Log.e(TAG, "[OTPTokenHandler]: RETURN FRE000 ERROR from onException");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.UNKNOWN_ERROR);
        }
    }

    public static void isValidAuthMethod(PluginCall call) {

        String baseUrl = call.getString("url");
        String trxId   = call.getString("trxId");
        JSObject payloadObj = call.getObject("payload");
        String payload = "{}";

        if (payloadObj != null) {
            payload = payloadObj.toString();
            Log.d(TAG, "PAYLOAD: " + payload);
        }

        String adviceXml = "<Advices><AttributeValuePair><Attribute name='TransactionConditionAdvice'/>"
                + "<Value>" + trxId + "</Value></AttributeValuePair></Advices>";
        String qs = "authIndexType=composite_advice&authIndexValue="
                + URLEncoder.encode(adviceXml, StandardCharsets.UTF_8);

        String url = baseUrl + (baseUrl.contains("?") ? "&" : "?") + qs;

        executeHttpQuery(url, payload, call);
    }

    private static void executeHttpQuery(String url, String payload, PluginCall call) {
        String sessionToken = FRSession.getCurrentSession().getSessionToken().getValue();
        String cookie = "iPlanetDirectoryPro=" + sessionToken;
        RequestBody body = RequestBody.create(payload, MEDIA_JSON);

        Request req = new Request.Builder()
                .url(url)
                .post(body)
                .addHeader("Content-Type", "application/json")
                .addHeader("Accept", "application/json")
                .addHeader("Accept-API-Version", "protocol=1.0,resource=2.0")
                .addHeader("Cookie", cookie)
                .build();

        handleHttpRequest(req, call);
    }

    private static void handleHttpRequest( Request req, PluginCall call){
        OkHttpClient client = httpClient();
        new Thread(() -> {
            try (Response resp = client.newCall(req).execute()) {

                String text = resp.body() != null ? resp.body().string() : "";
                Log.d(TAG, "RESPONSE"+ text);
                if (text.trim().isEmpty()) {
                    Log.e(TAG, "Without JSON");
                    return;
                }
                AuthMethodResponse(call, text);

            } catch (Exception e) {
                Log.e(TAG, "[OTPTokenHandler]: handleHttpRequest exception: " + e.getMessage(), e);
                ErrorHandler.reject(call, ErrorHandler.ErrorCode.HTTP_REQUEST_ERROR);
            }
        }).start();

    }

    private static void AuthMethodResponse(PluginCall call, String text) throws JSONException {
        JSONObject json = new JSONObject(text);

        String successResponse = json.optString("tokenId", null);
        String errorResponse = json.optString("code", null);
  

        if(successResponse != null || errorResponse != null){
            Log.d(TAG, "ENTRO AQUI: ");
            String status = successResponse != null ? "success": "failed";
            JSObject buildSuccessCallback =  buildSuccessErrorCallback(status);
            Log.d(TAG, "buildSuccessCallback r: " + buildSuccessCallback);
            call.resolve(buildSuccessCallback);
           
        }else{
            Log.d(TAG, "ENTRO ACA: ");
            Log.d(TAG, "JSON OK: " + json);
            call.resolve(JSObject.fromJSONObject(json));
        }

    }

    private static JSObject buildSuccessErrorCallback(String status) {
        try {

            JSObject outMessage = new JSObject()
                    .put("name", "status")
                    .put("value", status);

            JSArray output = (JSArray) new JSArray().put(outMessage);

            JSObject successCallback = new JSObject()
                    .put("type", "SuccessCallback")
                    .put("output", output);

            JSArray callbacks = (JSArray) new JSArray().put(successCallback);

            return new JSObject()
                    .put("callbacks", callbacks);
        } catch (Exception e) {
            Log.e(TAG, "SUCCESS ERROR "+e);
          throw new RuntimeException("Error building TextOutputCallback", e);
        }
      }

  private static JSObject validateExistMechanism(Context context) throws FRException {
      try{
        FRAClient fraClient = initClient(context);

        List<Account> accounts = fraClient.getAllAccounts();

        Mechanism found = null;

          if (accounts != null && !accounts.isEmpty()) {
              for (Account account : accounts) {
                  List<Mechanism> mechanisms = account.getMechanisms();
                  if (mechanisms != null && !mechanisms.isEmpty()) {
                      found = mechanisms.get(0);
                      break;
                  }
              }
          }

          JSObject result = new JSObject();
          result.put("empty",  accounts.isEmpty());

          if (found != null) {
              JSObject mechJson = new JSObject();
              mechJson.put("id", found.getId());
              mechJson.put("accountName", found.getAccountName());
              mechJson.put("issuer", found.getIssuer());
              result.put("mechanism", mechJson);
          } else {
              result.put("mechanism", null);
          }

          return result;
      } catch (Exception e) {
          Log.e(TAG, "error "+e);
          throw new FRException(ErrorHandler.ErrorCode.NO_ACCOUNTS_REGISTERED);
      }
    }

  private static long getRemainingTime(OathTokenCode token){
      long until = token.getUntil();
      long now = System.currentTimeMillis();

      return (until - now) / 1000;
    }

  private static FRAClient initClient(Context context) throws AuthenticatorException {
        return new FRAClient.FRAClientBuilder().withContext(context).start();
  }

  private static Account getAccount(FRAClient fraClient) throws FRException  {
        List<Account> accounts = fraClient.getAllAccounts();
        if (accounts == null || accounts.isEmpty()) {
          throw new FRException(ErrorHandler.ErrorCode.NO_ACCOUNTS_REGISTERED);
        }
        return accounts.get(0);
  }

   private static OathMechanism getOathMechanism(Account account) throws FRException  {
        for (Mechanism mechanism : account.getMechanisms()) {
            if (mechanism instanceof OathMechanism) {
                return (OathMechanism) mechanism;
            }
        }
      throw new FRException(ErrorHandler.ErrorCode.NO_OTP_REGISTERED);
    }

    private static void checkServerAndDeviceOtpState(PluginCall call, Context context, String uuid){
        String sessionToken = FRSession.getCurrentSession().getSessionToken().getValue();
        String cookie = "iPlanetDirectoryPro=" + sessionToken;
        Log.d(TAG, "sessionToken"+sessionToken);
        String base_url = call.getString("url");

        if (base_url == null) {
            Log.e(TAG, "[OTPTokenHandler]: Error in get url");
            ErrorHandler.reject(call, ErrorHandler.ErrorCode.MISSING_JOURNEY);
            return;
        }

        String url = base_url+"/"+uuid+"/devices/2fa/oath?_queryFilter=true";

        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder()
                .url(url)
                .addHeader("Cookie", cookie)
                .addHeader("Accept", "application/json")
                .get()
                .build();

        new Thread(() -> {
            try (Response response = client.newCall(request).execute()) {
                if (response.isSuccessful()) {

                    String json = response.body().string();

                    JSONObject jsonObject = new JSONObject(json);

                    Integer resultCount = jsonObject.getInt("resultCount");

                    boolean hasServerToken = !(resultCount == 0);
                    JSObject result = validateExistMechanism(context);
                    boolean hasDeviceToken = !result.getBool("empty");

                    if(hasDeviceToken && hasServerToken){
                        JSObject mechanism =result.getJSObject("mechanism");

                        hasServerToken = isOtpConsistentWithServer(uuid, mechanism);
                    }

                    Log.d(TAG, hasServerToken+ "hasDeviceToken"+hasDeviceToken);
                    sendOtpStatusResult(call, hasServerToken, hasDeviceToken);
                } else {
                    Log.e(TAG, "[OTPTokenHandler]:  Error HTTP: " + response.code());
                    ErrorHandler.reject(call, ErrorHandler.ErrorCode.HTTP_REQUEST_ERROR);
                }
            } catch (Exception e) {
                Log.e(TAG, "[OTPTokenHandler]:  Exception: " + e.getMessage(), e);
                ErrorHandler.reject(call, ErrorHandler.ErrorCode.HTTP_REQUEST_ERROR);
            }
        }).start();
    }

    private static boolean isOtpConsistentWithServer(String uuid , JSObject mechanism) {
         if(Objects.equals(uuid, mechanism.getString("accountName"))){
            return true;
        }
        return false;
    }

    private static void sendOtpStatusResult(PluginCall call, Boolean hasServerToken, Boolean hasDeviceToken ){
        JSObject result = new JSObject();

        result.put("hasServerToken", hasServerToken);
        result.put("hasDeviceToken", hasDeviceToken);
        Log.d(TAG, "result: " + result);
        call.resolve(result);
    }

}

File: ios/Sources/ForgerockBridgePlugin/OTPTokenHandler.swift
import Foundation
import Capacitor
import FRAuth
import FRCore
import FRAuthenticator

class OTPTokenHandler {
    private let call: CAPPluginCall
    private let plugin: ForgerockBridgePlugin
    private var fraClient: FRAClient? { FRAClient.shared }
    
    init(call: CAPPluginCall, plugin: ForgerockBridgePlugin) {
        self.call = call
        self.plugin = plugin
    }

    func startJourney(_ call: CAPPluginCall,completion: @escaping NodeCompletion<Token> ) {
        guard let journey = call.getString("journey"), !journey.isEmpty else {
            ErrorHandler.reject(call, code: ErrorCode.missingJourney)
            return
        }
        
          FRSession.authenticate(authIndexValue: journey, completion: completion)
    }

    
    func validateExistenceOTP(call: CAPPluginCall) {
        guard let user = FRUser.currentUser else {
            ErrorHandler.reject(call, code: ErrorCode.gettingUserInfo)
            return
        }

        user.getUserInfo { userInfo, error in
            if let error = error {
                ErrorHandler.reject(call, code: ErrorCode.gettingUserInfo)
            } else if let userInfo = userInfo {
                
                print("userInfo",userInfo.userInfo)
                
                guard let payload = userInfo.userInfo as? [String: Any] else {
                       ErrorHandler.reject(call, code: ErrorCode.gettingUserInfo)
                       return
                   }

                if let uuid = payload["sub"] as? String {
                    print("sub:", uuid)
                    self.checkServerAndDeviceOtpState(call: call, uuid: uuid );
                } else {
                   ErrorHandler.reject(call, code: ErrorCode.gettingUserInfo)
                }
            } else {
                ErrorHandler.reject(call, code: ErrorCode.gettingUserInfo)
            }
        }
    }
    
    func isValidAuthMethod(call: CAPPluginCall) {
      guard let baseUrl = call.getString("url"), !baseUrl.isEmpty else {
        call.reject("Missing 'url'")
        return
      }
    
     let trxId = call.getString("trxId") ?? ""

      var payloadJson = "{}"
      if let payloadObj = call.getObject("payload"),
         let data = try? JSONSerialization.data(withJSONObject: payloadObj, options: []),
         let jsonStr = String(data: data, encoding: .utf8) {
        payloadJson = jsonStr
      }

      let adviceXml = "<Advices><AttributeValuePair><Attribute name='TransactionConditionAdvice'/><Value>\(trxId)</Value></AttributeValuePair></Advices>"

      guard var comps = URLComponents(string: baseUrl) else {
        call.reject("Invalid 'url'")
        return
      }

      var items = comps.queryItems ?? []
      items.append(URLQueryItem(name: "authIndexType", value: "composite_advice"))
      items.append(URLQueryItem(name: "authIndexValue", value: adviceXml))
      comps.queryItems = items

      guard let finalUrl = comps.url?.absoluteString else {
        call.reject("Failed to build URL")
        return
      }

      executeHttpQuery(urlString: finalUrl, payload: payloadJson, call: call)
    }


    func executeHttpQuery(urlString: String, payload: String, call: CAPPluginCall){
        let sessionToken = FRSession.currentSession?.sessionToken?.value
        let cookie = "iPlanetDirectoryPro=\(sessionToken ?? "")"
         
        guard let url = URL(string: urlString) else {
            ErrorHandler.reject(call, code: .httpRequestError)
            return
       }
        
        print("PAYLOAD executeHttpQuery: \(payload)")
        var req = URLRequest(url: url)
        if let bodyData = payload.data(using: .utf8) {
            req.httpBody = bodyData
        }
            req.httpMethod = "POST"
            req.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
            req.setValue("application/json", forHTTPHeaderField: "Accept")
            req.setValue("protocol=1.0,resource=2.0", forHTTPHeaderField: "Accept-API-Version")
            req.setValue("XMLHttpRequest", forHTTPHeaderField: "X-Requested-With")
            req.setValue(cookie, forHTTPHeaderField: "Cookie")
            

           print("REQUEST: \(req)")
        
        
        self.handleHttpRequest(call: call, request: req)
   }
    
    func handleHttpRequest(call: CAPPluginCall, request: URLRequest) {
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            print("DATA: \(String(describing: data))")
            print("RESPONSE2: \(String(describing: response))")
            print("ERROR: \(String(describing: error))")
            
            
            if let error = error {
                print("[OTPTokenHandler] network error: \(error)")
                ErrorHandler.reject(call, code: .httpRequestError)
                return
            }
        
            guard let http = response as? HTTPURLResponse else {
                print("ERROR RESPONSE")
                ErrorHandler.reject(call, code: .httpRequestError)
                return
            }
        
            guard (200...299).contains(http.statusCode) else {
                print("HTTP error: \(http.statusCode) - \(HTTPURLResponse.localizedString(forStatusCode: http.statusCode))")
                ErrorHandler.reject(call, code: .httpRequestError)
                return
            }
            
            if let data = data {
                if let jsonString = String(data: data, encoding: .utf8) {
                    print("[Raw JSON Response]: \(jsonString)")
                }
                self.AuthMethodResponse(call: call, data)
            }
    
        }.resume()
        
    }
    
    func AuthMethodResponse (call: CAPPluginCall, _ data: Data) {
        if let jsonObject = (try? JSONSerialization.jsonObject(with: data, options: [])) as? [String: Any] {
            
            let successResponse = jsonObject["tokenId"] as? String ?? nil
            let errorResponse = jsonObject["code"] as? String ?? nil
            
            if(successResponse != nil || errorResponse != nil){
                
                let status = (successResponse != nil) ? "success" : "failed"
                print("[OTPTokenHandler]: Data send to front: ")
                print(jsonObject)
                self.buildSuccessErrorCallback(call: call, status: status)
                return
                
            }else {
                print("[OTPTokenHandler]: Json sent to front: ");
                call.resolve(jsonObject)
                return
            }
            
        }
        
    }
    
    func buildSuccessErrorCallback (call: CAPPluginCall, status: String) {
        
        let outMessage: [String: Any] = [
                "name": "status",
                "value": status
        ]
            
        let successCallback: [String: Any] = [
            "type": "SuccessCallback",
            "output": [outMessage]
        ]
        
        let result: [String: Any] = [
            "callbacks": [successCallback]
        ]
        
        call.resolve(result)
        
    }
    
    func checkServerAndDeviceOtpState(call: CAPPluginCall, uuid: String){
        
        let sessionToken = FRSession.currentSession?.sessionToken?.value
        let cookie = "iPlanetDirectoryPro=\(sessionToken ?? "")"
         
        guard let base_url = call.getString("url"), !base_url.isEmpty else {
            ErrorHandler.reject(call, code: ErrorCode.missingParameter)
            return
        }

        let urlString = "\(base_url)/\(uuid)/devices/2fa/oath?_queryFilter=true"
        guard let url = URL(string: urlString) else {
                ErrorHandler.reject(call, code: .httpRequestError)
                return
       }
        
        var req = URLRequest(url: url)
            req.httpMethod = "GET"
            req.setValue("application/json", forHTTPHeaderField: "Accept")
            req.setValue(cookie, forHTTPHeaderField: "Cookie")
        

        URLSession.shared.dataTask(with: req) { data, response, error in
                if let error = error {
                    print("[OTPTokenHandler] network error: \(error)")
                    ErrorHandler.reject(call, code: .httpRequestError)
                    return
                }
            
            guard let http = response as? HTTPURLResponse else {
                ErrorHandler.reject(call, code: .httpRequestError)
                return
            }
            
            guard (200...299).contains(http.statusCode) else {
                print("HTTP error: \(http.statusCode) - \(HTTPURLResponse.localizedString(forStatusCode: http.statusCode))")
                ErrorHandler.reject(call, code: .httpRequestError)
                return
            }

            guard let data = data else {
                print("Without data (data == nil)")
                ErrorHandler.reject(call, code: .httpRequestError)
                return
            }
            if let jsonString = String(data: data, encoding: .utf8) {
                print("[Raw JSON Response]: \(jsonString)")
            }

                do {
                   
                    struct DevicesResponse: Decodable {
                        let resultCount: Int?
                    }
                    let decoded = try JSONDecoder().decode(DevicesResponse.self, from: data)
                    let resultCount = decoded.resultCount ?? 0
                    var hasServerToken = !(resultCount == 0)

                    let mechanism = try self.validateExistMechanism();
                    let hasDeviceToken = !mechanism.empty;
                    
                    //If exist mechanism and token in server, has another validation. If mechanism and Server token are same.
                    //If not match = in this case hasDeviceToken always will be true, because in case of create new Token, the mechanism must be removed first and this response say to front thats exist mechanis and need delete first. If match same are true
                    if(hasDeviceToken && hasServerToken){
                        hasServerToken = self.isOtpConsistentWithServer(uuid: uuid, mechanism: mechanism.mechanism )
                    }
                  
                    call.resolve([
                        "hasServerToken": hasServerToken,
                        "hasDeviceToken": hasDeviceToken,

                    ])
                } catch {
 
                    NSLog("[OTPTokenHandler] JSON parse error: \(error)")
                    ErrorHandler.reject(call, code: .httpRequestError)
                }
            }.resume()

    }
    
    func isOtpConsistentWithServer(uuid: String, mechanism: Mechanism?) -> Bool{
        if let mechanism = mechanism {
            if(uuid ==  mechanism.accountName){
                return true;
            }
        }
        return false;
    }
    
    func hasRegisteredMechanism(_ call: CAPPluginCall) {
        do {
            let empty = try validateExistMechanism().empty

            var result = JSObject()
            result["empty"] = empty;
            call.resolve(result)
        } catch let error as ErrorCode {
            ErrorHandler.reject(call, code: error)
        } catch {
            ErrorHandler.reject(call, code: .unknown_error)
        }
    }
     func validateExistMechanism() throws -> (empty: Bool, mechanism: Mechanism?) {

                if let accounts = fraClient?.getAllAccounts() {
                    var empty = true;
                    var mechanism: Mechanism? = nil
                    for account in accounts {
                        if let first = account.mechanisms.first {
                           mechanism = first
                           empty = false
                           break
                        }
                    }
                    return (empty: empty, mechanism: mechanism)
                }else {
                    throw ErrorCode.noAccountsRegistered
                }
        }
    
    func generateOTP(call: CAPPluginCall) {
    do {
            guard let accounts = fraClient?.getAllAccounts(), !accounts.isEmpty else {
                throw ErrorCode.noAccountsRegistered
            }

             let token = try generateCode(from: accounts)

             var result = JSObject()
             result["otp"] = token.otp
             result["expiresIn"] = token.time
             call.resolve(result)
        
        } catch let error as ErrorCode {
            ErrorHandler.reject(call, code: error)
        } catch {
            ErrorHandler.reject(call, code: .unknown_error)
        }
    }


    func generateCode(from accounts: [Account]) throws -> (otp: String, time: Int)  {
        
        guard !accounts.isEmpty else {
           throw ErrorCode.noAccountsRegistered
        }
        
        for account in accounts {
            for mech in account.mechanisms {
                if let totpMechanism = mech as? TOTPMechanism {
                    if let otp = try? totpMechanism.generateCode() {
                        let remainingTime = getRemainingTime(from: otp);
                        return (otp: otp.code, time: remainingTime)
                    }
                }
            }
        }
        throw ErrorCode.noOtpRegistered
    }
    
    func getRemainingTime(from token: OathTokenCode) -> Int {
        guard let until = token.until else { return 0}
        let now = Date().timeIntervalSince1970
        return Int(until - now)
    }

  
}



